### ISSUE TITLE
Canvas Video Links?

### ISSUE THREAD

**Issue by itsmeow** [librespot-org/librespot#564]
[Spotify Canvas](https://canvas.spotify.com/) is a feature only available on some mobile versions of the app. Since librespot uses the same API, would it be possible to somehow get the canvas link for a given track? I've reverse engineered so much of it but come no closer to coming up with how to get the link.

Here's a link to an actual canvas and the location on the CDN:
https://canvaz.scdn.co/upload/artist/7aUSp5cOZlwEtd5zPC795k/video/74a85f3a67be4f69947f9f8c697d9d81.cnvs.mp4

**Comment by devgianlu** [comment]
Can't remember when I implement this. But you may want to have a look:
https://github.com/librespot-org/librespot-java/blob/ba576032510ac1fc1d91a8c9d7d48260d92cdd3b/lib/src/main/java/xyz/gianlu/librespot/dealer/ApiClient.java#L167-L175

**Comment by itsmeow** [comment]
Nice. Would it be possible to recreate a very rudimentary version of *only* this request with JavaScript? I have a stored session blob, and I really need it in JS. Is the protocol for the most part stateless or does Spotify have a lot of session-y token tricks? To be more specific, I'm trying to recreate this request in a Spicetify JS extension, so I have access to a running Spotify client, which means I probably already have some of this token stuff. Sorry if this is a bit outside your scope.

**Comment by itsmeow** [comment]
Nevermind. I figured it out. There's a client in Spotify / Spicetify's API called "Cosmos" that sends to those servers. If you send a request to `hm://canvaz-cache/v0/canvases`, and set the body to the protobuf byte array encoded as UTF-8 (messy, I know), it sends back a UTF-8 encoded byte array.
![image](https://user-images.githubusercontent.com/10366817/104911392-c90d0580-5958-11eb-80e5-d4cea56f5862.png)

```js
let payload = {
  url: `hm://canvaz-cache/v0/canvases`,
  body: `${new TextDecoder().decode(protobufEncodedBytes)}`,
};
Spicetify.CosmosAPI.resolver.post(payload, (error, res) => {
  if (error) {
    console.log(error);
    return;
  }
  let bytes = new TextEncoder().encode(res._body);
  let decoded = EntityCanvazResponse.decode(bytes);
  console.log(decoded);
});
```

Your code was really helpful for sorting the protobuf section, so thank you! I had to dig around in the client source a lot for the rest, though.

I think I've just made the first desktop based canvas implementation ever! I mean obviously right now all it does is grab the link, but the fact that it **WORKS** is amazing.
